### PR TITLE
GLT-2193: set completion date for failed Illumina runs (RunScanner)

### PR DESCRIPTION
**Pull request by djcooke**

### CODE REVIEW

File: runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java [hunk 1]
```diff
@@ -173,20 +178,28 @@ public NotificationDto process(File runDirectory, TimeZone tz) throws IOExceptio
     // The Illumina library can't distinguish between a failed run and one that either finished or is still going. Scan the logs, if
     // available to determine if the run failed.
     File rtaLogDir = new File(runDirectory, "/Data/RTALogs");
```
Comment (sqwci) [inline runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java]: MINOR Refactor your code to get this URI from a customizable parameter. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1075)

File: runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java [hunk 2]
```diff
@@ -173,20 +178,28 @@ public NotificationDto process(File runDirectory, TimeZone tz) throws IOExceptio
...
+              return null;
+            }
+            Matcher m = FAILED_MESSAGE.matcher(failMessage);
+            return LocalDateTime.parse(m.group(1), FAILED_MESSAGE_DATE_FORMATTER);
+          } catch (FileNotFoundException e) {
+            log.error("RTA file vanished before reading", e);
+            return null;
+          }
+        })
+        .filter(date -> date != null)
```
Comment (sqwci) [inline runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java]: MINOR Replace this lambda with method reference 'Objects::nonNull'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

**Comment by sqwci** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultIllumina.java#L142](https://github.com/TGAC/miso-lims/blob/1e89807a48d4e65f7791164296cc251e44f8f47e/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L142): Refactor this method to reduce its Cognitive Complexity from 33 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L83](https://github.com/TGAC/miso-lims/blob/1e89807a48d4e65f7791164296cc251e44f8f47e/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L83): String contains no format specifiers. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L155](https://github.com/TGAC/miso-lims/blob/1e89807a48d4e65f7791164296cc251e44f8f47e/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L155): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)
1. ![MINOR][MINOR] [DefaultIllumina.java#L79](https://github.com/TGAC/miso-lims/blob/1e89807a48d4e65f7791164296cc251e44f8f47e/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L79): Replace this "switch" statement by "if" statements to increase readability. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultIllumina.java#L143](https://github.com/TGAC/miso-lims/blob/897f6bc43e5bc21da54cec23744c7d0bb2a6ce60/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L143): Refactor this method to reduce its Cognitive Complexity from 33 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L84](https://github.com/TGAC/miso-lims/blob/897f6bc43e5bc21da54cec23744c7d0bb2a6ce60/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L84): String contains no format specifiers. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L156](https://github.com/TGAC/miso-lims/blob/897f6bc43e5bc21da54cec23744c7d0bb2a6ce60/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L156): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)
1. ![MINOR][MINOR] [DefaultIllumina.java#L80](https://github.com/TGAC/miso-lims/blob/897f6bc43e5bc21da54cec23744c7d0bb2a6ce60/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L80): Replace this "switch" statement by "if" statements to increase readability. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java [hunk 3]
```diff
@@ -173,20 +179,28 @@ public NotificationDto process(File runDirectory, TimeZone tz) throws IOExceptio
...
+            }
+            Matcher m = FAILED_MESSAGE.matcher(failMessage);
+            return LocalDateTime.parse(m.group(1), FAILED_MESSAGE_DATE_FORMATTER);
+          } catch (FileNotFoundException e) {
+            log.error("RTA file vanished before reading", e);
+            return null;
+          }
+        })
+        .filter(Objects::nonNull)
+        .findAny()
```
Comment (apmasell) [inline runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java]: I think you should `.sorted(LocalDateTime::compareTo).findFirst()` to get the earliest date.

**Pull request merged**